### PR TITLE
New methods `extractAll` and `extractSchema` to `UrlParams` (added `Schema.BooleanFromString`)

### PR DESCRIPTION
**Pull request by SandroMaglione**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Added `extractAll` and `extractSchema` to `UrlParams`. They allow to extract all key-value pairs from `UrlParams` and validate them using a given `Schema`.

```ts
const urlParams = UrlParams.fromInput({ a: 1, b: true, c: "string", e: [1, 2, 3] })
const result = UrlParams.extractAll(urlParams)
assert.deepStrictEqual(
  result,
  { "a": "1", "b": "true", "c": "string", "e": ["1", "2", "3"] }
)
```

```ts
Effect.gen(function*() {
  const urlParams = UrlParams.fromInput({ "a": [10, "string"], "b": false })
  const result = yield* UrlParams.extractSchema(Schema.Struct({
    a: Schema.Tuple(Schema.NumberFromString, Schema.String),
    b: Schema.BooleanFromString
  }))(urlParams)

  assert.deepStrictEqual(result, {
    a: [10, "string"],
    b: false
  })
}))
```

The PR also added a new `Schema.BooleanFromString` to `Schema`.

## Motivation
When using `searchParams` from frameworks like `next`, I often need to validate the correct shape using a `Schema`. The params have this type:

```ts
type SearchParamsNext = Promise<{ [key: string]: string | string[] | undefined }>;
```

Using `fromInput` + `extractSchema` allows to make sure that the params are valid (or handle a `ParseError` otherwise).

### CODE REVIEW

File: packages/effect/src/Schema.ts [hunk 1]
```diff
@@ -9245,6 +9245,19 @@ export class BooleanFromUnknown extends transform(
   { strict: true, decode: Predicate.isTruthy, encode: identity }
 ).annotations({ identifier: "BooleanFromUnknown" }) {}
 
+/**
+ * Converts an `string` value into its corresponding `boolean`
+ * ("true" as `true` and "false" as `false`).
+ *
+ * @category boolean constructors
```
Comment (gcanti) [inline packages/effect/src/Schema.ts:9252]: should be "boolean transformations"

File: packages/effect/src/Schema.ts [hunk 2]
```diff
@@ -9245,6 +9245,19 @@ export class BooleanFromUnknown extends transform(
   { strict: true, decode: Predicate.isTruthy, encode: identity }
 ).annotations({ identifier: "BooleanFromUnknown" }) {}
 
+/**
+ * Converts an `string` value into its corresponding `boolean`
+ * ("true" as `true` and "false" as `false`).
```
[the remark this answers is not in the document]
Comment (gcanti) [reply packages/effect/src/Schema.ts:9250]:
```ts
export class BooleanFromString extends transform(
  Literal("true", "false"),
  Boolean$,
  { strict: true, decode: (value) => value === "true", encode: (value) => value ? "true" : "false" }
).annotations({ identifier: "BooleanFromString" }) {}
```

[the remark this answers is not in the document]
Comment (SandroMaglione) [reply packages/effect/src/Schema.ts:9250]: Is there a more extensive way to test this beside what it's [already in the PR](https://github.com/Effect-TS/effect/blob/2a9b16cf55d5f9a5c32b03373eb30c9b5ccfcbca/packages/effect/test/Schema/Schema/Boolean/BooleanFromString.test.ts)?

[the remark this answers is not in the document]
Comment (gcanti) [reply packages/effect/src/Schema.ts:9250]: you can add a test to verify a decoding failure, such as:

```ts
    await Util.expectDecodeUnknownFailure(
      schema,
      "a",
      `BooleanFromString
└─ Encoded side transformation failure
   └─ "true" | "false"
      ├─ Expected "true", actual "a"
      └─ Expected "false", actual "a"`
    )
```

**Review by tim-smart — APPROVED** [review]

**Pull request merged**